### PR TITLE
Fix dialyzer issues

### PR DESCRIPTION
**Pull request by davydog187**
Fixes nearly two dozen dialyzer issues. Need some help with the proper typespecs for a few

### CODE REVIEW

**Comment by davydog187** [comment]
FYI that `rebar3 dialyzer` passes on master, while running it from `mix dialyzer` does not pass

File: .dialyzer_ignore.exs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[
+  {"lib/open_telemetry/span.ex", :invalid_contract}
```
Comment (davydog187) [inline .dialyzer_ignore.exs:2]: technically all of the span macros are invalid as they return AST. I think this is the only way to solve it unless we want to remove the typespecs

File: lib/open_telemetry/tracer.ex [hunk 2]
```diff
@@ -20,7 +20,8 @@ defmodule OpenTelemetry.Tracer do
   @type start_opts() :: %{
           optional(:parent) => OpenTelemetry.span() | OpenTelemetry.span_ctx(),
           optional(:attributes) => OpenTelemetry.attributes(),
-          optional(:sampler) => :ot_sampler.sampler(),
+          # TODO sampler should is an opaque type defined in the implementation
+          optional(:sampler) => term(),
```
Comment (davydog187) [inline lib/open_telemetry/tracer.ex:24]: I'm not sure how to resolve this one. The type is defined here https://github.com/open-telemetry/opentelemetry-erlang/blob/master/src/ot_sampler.erl#L35 which is not a dependency of this library

File: src/opentelemetry.erl [hunk 3]
```diff
@@ -119,8 +118,6 @@
 %% Note that this also excludes tabs, newlines, carriage returns, etc.
 -type tracestate()         :: [{unicode:latin1_chardata(), unicode:latin1_chardata()}].
 
--type stack_trace()        :: [erlang:stack_item()].
```
Comment (davydog187) [inline src/opentelemetry.erl:122]: these appeared unused, and are also being deprecated in OTP 24

**Comment by davydog187** [comment]
I need to figure out the github actions caching here. Also, for some reason `persistent_term:put/2` is failing on OTP 21.3 even though it should be available 🤔

**Comment by davydog187** [comment]
Looks like absinthe experienced the same issue

https://elixirforum.com/t/help-with-dialyzer-and-nowarn-function/30246
https://github.com/absinthe-graphql/absinthe/pull/895/commits/445e662ca92b789c23555fe9dcb3b7307f340756

File: .dialyzer_ignore.exs (hunk 1 above)
Comment (tsloughter) [reply .dialyzer_ignore.exs:2]: Hm, why is that invalid? Isn't that how they should work?

Comment (davydog187) [reply .dialyzer_ignore.exs:2]: Technically a macro is always returning AST, so dialyzer will not accept that we annotate it with a spec that says it returns a boolean

File: src/opentelemetry.erl (hunk 3 above)
Comment (tsloughter) [reply src/opentelemetry.erl:122]: Weird, what is it being replaced with in OTP 24?

**Comment by tsloughter** [comment]
Huh, so the `persistent_term:put` only with Elixir's dialyzer call... that is really odd.

**Comment by tsloughter** [comment]
@davydog187 any update on this?

**Comment by bryannaegele** [comment]
I think we just need to add a dialyzer ignore file

**Comment by tsloughter** [comment]
So elixir/mix doesn't have support for ignoring dialyzer warnings either in the module or in the mix config?

**Comment by davydog187** [comment]
We've been working on a launch the last few weeks. I'll try to pick this back up during this week

**Comment by davydog187** [comment]
Regarding suppressing the dialyzer warnings: that can be done through the .dialyzer_ignore.exs file.

I'm not 💯 on this, but that won't help dependent code.

File: .dialyzer_ignore.exs (hunk 1 above)
Comment (tsloughter) [reply .dialyzer_ignore.exs:2]: Oh, duh :)

**Comment by tsloughter** [comment]
I'll try moving these changes over to the new api under the opentelemetry-erlang repo https://github.com/open-telemetry/opentelemetry-erlang/tree/master/apps/opentelemetry_api

**Comment by hauleth** [comment]
@tsloughter `git subtree pull -P apps/opentelemetry_api repo branch`

**Comment by tsloughter** [comment]
@hauleth fails saying it is an "unrelated history", I'm guessing because of the PR having been squashed and merged.

**Comment by hauleth** [comment]
Well, it seems so. In that case I do not have simple way to port other than generating diff and manually applying it.

I can do it in a moment.

**Comment by davydog187** [comment]
I can just close and reopen it tomorrow. Would that be better for you guys?

**Comment by tsloughter** [comment]
I can let you know tomorrow morning, I'll see if I can knock it out in short order today.

**Comment by hauleth** [comment]
I close this, as it is no longer relevant as it was ported to the open-telemetry/opentelemetry-erlang#101 and merged there.

**Pull request closed without merging**